File: app/algotrading/api/schemas/deployment.py
"""Request and response schemas for pre-deployment model selection."""

from __future__ import annotations

from datetime import datetime
from enum import Enum

from algotrading.api.schemas.backtesting import PerformanceMetrics
from pydantic import BaseModel, Field


class ReadinessCheckStatus(str, Enum):
    """Possible outcomes for one deployment readiness criterion."""

    PASSED = "passed"
    WARNING = "warning"
    FAILED = "failed"


class ReadinessCheck(BaseModel):
    """One readiness criterion returned by deployment validation."""

    key: str
    label: str
    status: ReadinessCheckStatus
    message: str


class DeploymentReadiness(BaseModel):
    """Deployability verdict for a selected model generation."""

    deployable: bool
    errors: list[str] = Field(default_factory=list)
    warnings: list[str] = Field(default_factory=list)
    checks: list[ReadinessCheck] = Field(default_factory=list)


class DeploymentValidationRequest(BaseModel):
    """Payload used to validate a deployment candidate selection."""

    model_id: str = Field(min_length=1)
    generation_id: str = Field(min_length=1)


class DeploymentBacktestSummary(BaseModel):
    """Latest evaluation/backtest evidence for one generation."""

    source: str
    backtest_id: str | None = None
    completed_at: datetime | None = None
    metrics: PerformanceMetrics | dict[str, float | int | None]


class DeploymentGenerationSummary(BaseModel):
    """Generation metadata displayed by the deployment workflow."""

    generation_id: str
    generation_number: int
    status: str
    created_at: datetime
    training_duration_seconds: float
    final_reward: float | None = None
    model_path: str | None = None
    metrics: dict[str, float | int | str | bool | None] = Field(default_factory=dict)
    evaluation: DeploymentBacktestSummary | None = None


class DeploymentCandidate(BaseModel):
    """Core RL model candidate plus generation and readiness metadata."""

    model_id: str
    name: str
    description: str | None = None
    algorithm: str
    trainer_type: str
    state: str
    supporting_model_ids: list[str] = Field(default_factory=list)
    strategy_ids: list[str] = Field(default_factory=list)
    available_generations: list[DeploymentGenerationSummary] = Field(
        default_factory=list
    )
    latest_backtest: DeploymentBacktestSummary | None = None
    readiness: DeploymentReadiness


class DeploymentSelection(BaseModel):
    """Persisted pre-deployment candidate selection consumed by later phases."""

    model_id: str
    generation_id: str
    selected_at: datetime
    candidate: DeploymentCandidate
    readiness: DeploymentReadiness
    selected_by: str | None = None


__all__ = [
    "ReadinessCheckStatus",
    "ReadinessCheck",
    "DeploymentReadiness",
    "DeploymentValidationRequest",
    "DeploymentBacktestSummary",
    "DeploymentGenerationSummary",
    "DeploymentCandidate",
    "DeploymentSelection",
]

File: frontend/src/api/deployment.ts
import { apiClient } from './client';
import type { PerformanceMetrics } from './backtesting';

export type ReadinessCheckStatus = 'passed' | 'warning' | 'failed';

export interface ReadinessCheck {
  key: string;
  label: string;
  status: ReadinessCheckStatus;
  message: string;
}

export interface DeploymentReadiness {
  deployable: boolean;
  errors: string[];
  warnings: string[];
  checks: ReadinessCheck[];
}

export interface DeploymentValidationRequest {
  model_id: string;
  generation_id: string;
}

export interface DeploymentBacktestSummary {
  source: string;
  backtest_id?: string | null;
  completed_at?: string | null;
  metrics: PerformanceMetrics | Record<string, number | null>;
}

export interface DeploymentGenerationSummary {
  generation_id: string;
  generation_number: number;
  status: string;
  created_at: string;
  training_duration_seconds: number;
  final_reward?: number | null;
  model_path?: string | null;
  metrics: Record<string, number | string | boolean | null | undefined>;
  evaluation?: DeploymentBacktestSummary | null;
}

export interface DeploymentCandidate {
  model_id: string;
  name: string;
  description?: string | null;
  algorithm: string;
  trainer_type: string;
  state: string;
  supporting_model_ids: string[];
  strategy_ids: string[];
  available_generations: DeploymentGenerationSummary[];
  latest_backtest?: DeploymentBacktestSummary | null;
  readiness: DeploymentReadiness;
}

export interface DeploymentSelection {
  model_id: string;
  generation_id: string;
  selected_at: string;
  candidate: DeploymentCandidate;
  readiness: DeploymentReadiness;
  selected_by?: string | null;
}

/** API helpers for pre-deployment candidate selection. */
export const deploymentApi = {
  listCandidates(): Promise<DeploymentCandidate[]> {
    return apiClient.get<DeploymentCandidate[]>('/deployment/candidates');
  },

  validate(request: DeploymentValidationRequest): Promise<DeploymentReadiness> {
    return apiClient.post<DeploymentReadiness>('/deployment/validate', request);
  },

  getSelection(): Promise<DeploymentSelection | null> {
    return apiClient.get<DeploymentSelection | null>('/deployment/selection');
  },

  saveSelection(request: DeploymentValidationRequest): Promise<DeploymentSelection> {
    return apiClient.put<DeploymentSelection>('/deployment/selection', request);
  },

  clearSelection(): Promise<{ status: string }> {
    return apiClient.delete<{ status: string }>('/deployment/selection');
  },
};